org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/codeactions/LSPCodeActionQuickAssistProcessor.java

```diff
@@ -138,19 +132,18 @@ public boolean canAssist(IQuickAssistInvocationContext invocationContext) {
 	@Override
 	public ICompletionProposal[] computeQuickAssistProposals(IQuickAssistInvocationContext invocationContext) {
 		if (this.infos == null || this.infos.isEmpty()) {
-			infos = getLSPDocumentInfos();
+			infos = getLSPDocumentInfos(invocationContext.getSourceViewer().getDocument());
 			if (infos.isEmpty()) {
 				return NO_PROPOSALS;
 			}
 		}
 
 		// Get the codeActions
-		CodeActionParams params = prepareCodeActionParams(this.infos);
+		CodeActionParams params = prepareCodeActionParams(infos, invocationContext.getOffset(), invocationContext.getLength());
+		List<ICompletionProposal> proposals = Collections.synchronizedList(new ArrayList<>());
 		List<CompletableFuture<Void>> futures = Collections.emptyList();
 		try {
 			// Prevent infinite recursion by only computing proposals if there aren't any
-			if (proposals.contains(COMPUTING) || proposals.isEmpty()) {
```

Comment:
Why does it have to be removed? IIRC the case described in comment above still applies.

Reply:
I didn't understand the comment to be honest. I don't think this method can make recursive calls... I think it can be re-entrant (maybe the comment was referring to re-entrance) and hence there was proposals caching etc. but i didn't see the benefit (might be missing/misunderstanding) of this caching...

Reply:
I'm not sure this comment is clear enough, but it could be that the same LSPCodeActionQuickAssistProcessor can be invoked in several threads and that we don't want it to always recompute if there is no need for it. If this guard doesn't cause any problem for your usage, I suggest you keep it and we consider removing it in a separate bug.

Reply:
I have finally understood why proposals were cached. If proposals computation takes over 200ms then once they are computed the same method is invoked later (re-entrance) with the same invocation context object. Proposals need to be wiped out if context object is the same as before. I have added comments in the code.